registry/registry.go

```diff
@@ -161,400 +171,435 @@ func (r *factory) CreateEventRegistry(meta *types.Metadata) (EventRegistry, erro
+		return &ValueDecoder[string]{}, nil
 	case types.IsU8:
-		return &PrimitiveFieldType[types.U8]{}, nil
+		return &ValueDecoder[types.U8]{}, nil
 	case types.IsU16:
-		return &PrimitiveFieldType[types.U16]{}, nil
+		return &ValueDecoder[types.U16]{}, nil
 	case types.IsU32:
-		return &PrimitiveFieldType[types.U32]{}, nil
+		return &ValueDecoder[types.U32]{}, nil
 	case types.IsU64:
-		return &PrimitiveFieldType[types.U64]{}, nil
+		return &ValueDecoder[types.U64]{}, nil
 	case types.IsU128:
-		return &PrimitiveFieldType[types.U128]{}, nil
+		return &ValueDecoder[types.U128]{}, nil
 	case types.IsU256:
-		return &PrimitiveFieldType[types.U256]{}, nil
+		return &ValueDecoder[types.U256]{}, nil
 	case types.IsI8:
-		return &PrimitiveFieldType[types.I8]{}, nil
+		return &ValueDecoder[types.I8]{}, nil
 	case types.IsI16:
-		return &PrimitiveFieldType[types.I16]{}, nil
+		return &ValueDecoder[types.I16]{}, nil
 	case types.IsI32:
-		return &PrimitiveFieldType[types.I32]{}, nil
+		return &ValueDecoder[types.I32]{}, nil
 	case types.IsI64:
-		return &PrimitiveFieldType[types.I64]{}, nil
+		return &ValueDecoder[types.I64]{}, nil
 	case types.IsI128:
-		return &PrimitiveFieldType[types.I128]{}, nil
+		return &ValueDecoder[types.I128]{}, nil
 	case types.IsI256:
-		return &PrimitiveFieldType[types.I256]{}, nil
+		return &ValueDecoder[types.I256]{}, nil
 	default:
-		return nil, fmt.Errorf("unsupported primitive type %v", primitiveTypeDef)
+		return nil, ErrPrimitiveTypeNotSupported.WithMsg("primitive type %v", primitiveTypeDef)
 	}
 }
 
-// getStoredFieldType will attempt to return a field type from storage, and perform an extra check for recursive types.
-func (r *factory) getStoredFieldType(fieldName string, fieldType int64) (FieldType, bool) {
-	if ft, ok := r.fieldStorage[fieldType]; ok {
-		if rt, ok := ft.(*RecursiveFieldType); ok {
-			r.recursiveFieldStorage[fieldType] = rt
+// getStoredFieldDecoder will attempt to return a FieldDecoder from storage,
+// and perform an extra check for recursive decoders.
+func (f *factory) getStoredFieldDecoder(fieldLookupIndex int64) (FieldDecoder, bool) {
+	if ft, ok := f.fieldStorage[fieldLookupIndex]; ok {
+		if rt, ok := ft.(*RecursiveDecoder); ok {
+			f.recursiveFieldStorage[fieldLookupIndex] = rt
 		}
 
 		return ft, ok
 	}
 
 	// Ensure that a recursive type such as Xcm::TransferReserveAsset does not cause an infinite loop
-	// by adding the RecursiveFieldType the first time the field is encountered.
-	r.fieldStorage[fieldType] = &RecursiveFieldType{
-		FieldName: fieldName,
-	}
+	// by adding the RecursiveDecoder the first time the field is encountered.
+	f.fieldStorage[fieldLookupIndex] = &RecursiveDecoder{}
 
 	return nil, false
 }
 
-func (r *factory) storeFieldType(fieldType int64, registryFieldType FieldType) {
-	r.fieldStorage[fieldType] = registryFieldType
-}
-
 const (
-	unknownFieldName = "unknown_field_name"
-
-	fieldPathSeparator = "::"
+	fieldSeparator    = "_"
```

Comment:
Some substrate event/properties are named with underscores, so if we use `_` as field separator, then we will not be able to programatically recover the original key. Although usually the use of underscore is on the last property name and not on the package hierarchy above, so we can circumvent this issue by assuming that. 
Another reason to maybe use `.` instead is to be consistent with event names or Calls where it is used. 
WDYT?

Reply:
No particular feeling about this, we can use whatever from my point of view. Although, I do think we should only use one separator to make things easier(?).